### ISSUE TITLE
Sign requests using NewRequestAuth and sign URLs using NewURLAuth instead of using NewURLAuth for both

### ISSUE THREAD

**Issue by petemoore** [taskcluster/taskcluster-cli#83]
@jonasfj 

[This](https://github.com/taskcluster/taskcluster-cli/blob/7f91f2b09154aa2dbc7fe0810af702c23c8408f2/client/credentials.go#L102) can't work because [here](https://github.com/taskcluster/taskcluster-cli/blob/7f91f2b09154aa2dbc7fe0810af702c23c8408f2/client/credentials.go#L62) we are generating URL auth, and [URL Auth doesn't set a nonce](https://github.com/tent/hawk-go/blob/d3073d048f26babfe9c0b12c0e333ca71c13213a/hawk.go#L293-L297) (since [bewits are intentionally not protected from replay attack](https://github.com/hueniverse/hawk#bewit-limitations), hence why they should only be used for GET requests).

Instead, we need to use [Request auth](https://github.com/tent/hawk-go/blob/d3073d048f26babfe9c0b12c0e333ca71c13213a/hawk.go#L274-L284) for signing a request, not URL auth... :-)

Otherwise, it would probably go unnoticed (client may operate without problems) but we'd have a gaping security hole...

**Comment by djmitche** [comment]
Just to clarify since I'm not sure I understand:  This is an issue with `Credentials.SignHeader` which is ued by `Credentials.SignRequest`.  So this would affect any authenticated API call!  #78 shows that those have not historically worked.  Perhaps even with that landed, they still do not work?

This would be a good one to fix -- and to add a test for!

**Comment by jonasfj** [comment]
Yeah, I don't think these have worked... Some of the existing code is more like pseudo code :)

**Comment by djmitche** [comment]
Can you provide a way to reproduce this issue?

**Comment by jonasfj** [comment]
I suspect trying to use it will reproduce the error...
The signurl stuff was added because that's how we download artifacts...

Den 30. jan. 2017 2.41 PM skrev "Dustin J. Mitchell" <
notifications@github.com>:

Can you provide a way to reproduce this issue?

—
You are receiving this because you were assigned.

Reply to this email directly, view it on GitHub
<https://github.com/taskcluster/taskcluster-cli/issues/83#issuecomment-276215462>,
or mute the thread

**Comment by jonasfj** [comment]
I think this bug sums up to: auth stuff in tc-cli is likely broken.
Use existing code as pseudo code for a new implementation, and read the docs on how authentication works.

Note: node hawk library is the reference implementation for hawk, our docs adds a few extra concepts.

**Comment by djmitche** [comment]
```
./taskcluster awsProvisioner workerType -- tutorial
```
works for me, so I think that basic authentication stuff is OK, now that #78 is complete.  So perhaps it's just signed URLs that don't work.

@petemoore can you provide a little more context here?  Maybe there's just some dead code we can eliminate?

**Comment by djmitche** [comment]
It's worth noting that we should be using taskcluster/taskcluster-client-go to generate signed URLs, not doing so internally..

**Comment by jonasfj** [comment]
using taskcluster/taskcluster-client-go isn't easy because it's so heavily typed... So you can't do that when operating on reference JSON files... Ie. when dynamically interpreting arguments based on the contents of a reference file.

**Comment by jonasfj** [comment]
@djmitche, so if yes, downloading an artifact we can use taskcluster/taskcluster-client-go.

But if we're adding a command option like `--sign-url` that would work like: `taskcluster api queue --sign-url task <taskId>`
and return a signed url, we can't use the golang client, as it's heavily typed.

**Comment by djmitche** [comment]
Right, OK.  So there are two issues here:
 * API methods have no way to generate a signed URL for an operation, rather than executing it (no `--sign-url` option)
 * If that option existed, it wouldn't work because of the bug @petemoore has identified.

So I'm going to change the title of this issue here so that we both add that new feature, and convince it to work properly.

The expectation is that any API endpoint that uses the "GET" method can be called like this:

```
taskcluster queue getArtifact --sign-url taskid 0 public/live_backing.log
```

should simply output a URL to stdout.  Users might use this in a shell script, passing the resulting URL to `curl` or `wget` or even sending it to some other service that needs to download the artifact.

**Comment by petemoore** [comment]
Hey guys, I'm back! :-)

So the issue I was raising here is that hawk provides two different auth "profiles" for signing HTTP requests - one for signing the URL, which rewrites the URL path (i.e. bewits) and one for signing a request (leaving URL path intact, but setting an Authorization HTTP header). The main difference is that the auth "profile" used for bewits does not provide replay protection, since it does not set a nonce. This means a signed URL can be used multiple times, and it will consistently retrieve a given resource. In contrast, a signed URL (with an Authorization header) can only be used once, because the server will store previously used nonce values, and not allow the same nonce to be used for the same request.

The problem with the current implementation is that both URL signing (URL path rewriting) and request signing (Setting Authorization header) are using the same auth profile - the profile that is intended only for URL signing. In practical terms this means that the Authorization header created does not include a nonce, which means any signed request can be replayed, without replay protection.

In short, the fix is to not use NewURLAuth for both types of signing, but use NewRequestAuth for request signing, and NewURLAuth for URL signing.

I can also make a PR for this, I probably should have done that in the first place! :-)

**Comment by petemoore** [comment]
So  .... I went to make the PR - and realised this was already implemented an alternative way in 62574d8dc4cd234d6182e13b8cb3e6fad594e392 - whoops! So this is probably solved. What threw me is that rather than us calling `NewRequestAuth` and `NewURLAuth`, we still exclusively call `NewURLAuth`, but then effectively apply-the-diff ourselves by manipulating the returned Auth object [here](https://github.com/taskcluster/taskcluster-cli/blob/43df14d844e8b3996bd21f24fbc8ca1283b3ba99/client/credentials.go#L104). This feels a bit hacky (rather than calling correct methods upfront) but probably works.

Sorry for the noise!
